designs/2022-forbid-disable/README.md

```diff
@@ -0,0 +1,93 @@
+
+I will add the option to pass `noInlineConfig` an array of rule names. This is in addition to the already supported boolean setting. This will result in those rules having inline configuration disabled.
+
+This means that I will need to modify `config-schema.js`
+```
+noInlineConfig: { type: ["boolean", "array"] },
+```
+
+EG `// eslint-disable semi` will have ~~semi~~ as an error, as well as the violation that it is attempting to disable also having an error.
+
+The error on the inline config will have a severity of 2. See [LintMessage#severity](https://eslint.org/docs/developer-guide/nodejs-api#-lintmessage-type)
+
+first in `getDirectiveComments` I will match any directives that affect a rule listed in the array `noInlineConfig`. If there is a match, I will create a problem on that line to alert the user that they are attempting to disable a rule that will not be affected by the inline config.
+
+This error will **not** have a quick fix available
+
+
+In `applyDirectives` (`apply-disable-directives.js`) I will reference the problem. If that problem matches a name in the array `noInlineConfig` I will not apply any directives. This will ensure that `problems.push(problem)` gets run. Note that this same logic applies to configuration comments as well.
+
+## Documentation
+
+We will want to update the documentation ([here](https://eslint.org/docs/user-guide/configuring/rules#disabling-inline-comments)) to reflect the new changes. The announcement can be integrated into the config update announcement.
+
+## Drawbacks
+
+this adds complexity to the already complex config experience.
+
+This will have a very minor impact on perf.
+
+This design does not alert users in the case where they use `// eslint-disable` because that directive has `ruleId: null`. However, the design will still appropriately enforce the proper rules as desired.
+
+## Backwards Compatibility Analysis
+
+This should integrate into the planned config changes and therefore will not be backwards compatible. However, this is acceptable since that will be a breaking change that is already approved and planned.
+
+## Alternatives
+
+instead of extending `noInlineConfig` we could instead add a state called `enforce` to `error/warn`. This would be a superset of `error` where the rules cannot be disabled by inline config. However, it is hard to convey to the user succinctly what this accomplishes. And it allows users to disable the rules more easily in config files
+
+There are plugins such as [eslint-comments/no-restricted-disable](https://mysticatea.github.io/eslint-plugin-eslint-comments/rules/no-restricted-disable.html) that implement this concept. However, since these are just other rules, one can easily disable them. This results in scenarios such as
+
+```
+// eslint-disable-next-line eslint-comments/no-restricted-disable
+// eslint-disable no-undef
+```
+In large PRs, reviewers often miss these comments. In a large repo that I work on these scenario has happened over 30 times
```

Comment:
The problem with these rules is that they can be disabled. And users often just autofix these issues by disabling more. In the repos that I work on we have rules very similar to these. And they just get disabled. Like stacking dolls we have entire blocks of disable comments. For what it's worth, I would only configure ESLint to restrict inline config on rules like the ones above. To force compliance

Reply:
So perhaps implementing native replacements for each of these rules would satisfy your needs then (instead of relying on 3rd-party disable-able rules for them).

Reply:
Those proposed native replacements themselves can be disabled. Which will happen as evidence below from a repo I work on (this block can be found over 30 times in different contexts)

```
// eslint-disable-next-line foo/no-abusive-eslint-disable
// eslint-disable-next-line foo/no-eslint-disable
// eslint-disable-next-line foo/no-istanbul-disable-with-exceptions
// istanbul ignore file
```